hybrid/src/main/java/jetbrains/jetpad/hybrid/PrettyHybridProperty.java

```diff
@@ -177,53 +217,55 @@ public void run() {
     protected void afterItemRemoved(final int index, Token item, boolean success) {
       if (success) {
         updateValue(parse());
-        myRemoveHandler.handle(index);
         inSync(new Runnable() {
           @Override
           public void run() {
+            myRemoveHandler.handle(index);
             if (myValue != null) {
-              updateFromPretty();
+              updateFromPretty(index);
             } else {
-              mySource.remove(index);
+              sourceRemove(index);
             }
           }
         });
       }
     }
 
-    private void inSync(Runnable r) {
-      if (!myInSync) {
-        myInSync = true;
-        try {
-          r.run();
-        } finally {
-          myInSync = false;
-        }
-      }
-    }
-
-    private void updateFromPretty() {
+    private void updateFromPretty(final int forceToSource) {
       PrettyPrinterContext<? super ModelT> printCtx = new PrettyPrinterContext<>(myPrinter);
       printCtx.print(myValue);
       List<Token> prettyTokens = printCtx.tokens();
+      boolean forceUpdated = false;
       for (int i = 0; i < prettyTokens.size(); i++) {
-        Token p = prettyTokens.get(i);
+        final Token p = prettyTokens.get(i);
         if (i < myTokens.size() && i < mySource.size()) {
           // Token exists in both source and tokens.
           if (!Objects.equals(p, myTokens.get(i))) {
             myTokens.set(i, p);
-            mySource.set(i, myFrom.apply(i, p));
+            if (i == forceToSource) {
+              forceUpdated = true;
+            }
+            sourceSet(i, p);
           }
         } else if (i < myTokens.size() && i >= mySource.size()) {
           // Token just appeared in tokens and doesn't exists in source.
           if (!Objects.equals(p, myTokens.get(i))) {
             myTokens.set(i, p);
           }
-          mySource.add(i, myFrom.apply(i, p));
+          if (i == forceToSource) {
+            forceUpdated = true;
+          }
+          sourceAdd(i, p);
         } else {
-          // A new token added.
-          myTokens.add(p);
-          mySource.add(myFrom.apply(i, p));
+          throw new IllegalStateException("A printer generated more tokens than where is source.");
```

Comment:
I would use this text:
"A printer generated more tokens than there are in the source."
or 
"A printer generated more tokens than the source."